runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcGetDataStream.java

```diff
@@ -302,38 +331,57 @@ public void appendSpecificHtml(PrintWriter writer) {
   }
 
   private <ResponseT> ResponseT issueRequest(QueuedRequest request, ParseFn<ResponseT> parseFn) {
-    while (true) {
+    while (!isShutdown()) {
       request.resetResponseStream();
       try {
         queueRequestAndWait(request);
         return parseFn.parse(request.getResponseStream());
-      } catch (CancellationException e) {
-        // Retry issuing the request since the response stream was cancelled.
-        continue;
+      } catch (AppendableInputStream.InvalidInputStreamStateException
+          | VerifyException
```

Comment:
throughout the class when we insert/remove into the pending map we call `verify(map operation)`
so that might get thrown here

Reply:
If it is thrown and we are shutdown, the verify exception is invalid because any ongoing responses/requests are invalid.  We clear the `pending` + `requests` fields so some of the map operations are looking for things that are no longer there.

An alternative is instead of just verify, we add a check for isShutdown ontop of the boolean condition being checked and either return or throw if isShutdown.  The issue here is we start checking for isShutdown every vs letting the exception just get thrown and handled here.

If the stream is not shutdown, the VerifyException will still be thrown.  I can add a test for this?

Reply:
If we see VerifyException, that means there is a logic bug and we need to correct code to fix them. 

One case I see is, on shutdown we do `batches.clear()`, this could happen in the middle of the batch sending. There is `verify(batch == batches.peekFirst());` inside `send` that will throw after shutdown.

from https://github.com/google/guava/wiki/ConditionalFailuresExplained#kinds-of-conditional-failures
```
A verification check is used when you lack high confidence that an API you consume will meet its (real or implied) specification. It's easiest to understand this type of check as "like an assertion in almost every way, but we'll never want to disable them."
``` 

Can we make sure Verify exceptions are not thrown unless there is a bug?

Reply:
added checks to isShutdown